avalanche/benchmarks/utils/flat_data.py

```diff
@@ -44,6 +44,66 @@
 T_co = TypeVar("T_co", covariant=True)
 
 
+class LazyIndices:
+    """More efficient ops for indices.
+
+    Avoid wasteful allocations, accept generators. Convert to list only
+    when needed.
+
+    Do not use for anything outside this file.
+    """
+
+    def __init__(self, *lists, known_length=None, offset=0):
+        new_lists = []
+        for ll in lists:
+            if isinstance(ll, LazyIndices) and ll._eager_list is not None:
+                # already eagerized, don't waste work
+                new_lists.append(ll._eager_list)
+            else:
+                new_lists.append(ll)
+        self._lists = new_lists
+
+        if len(self._lists) == 1 and offset == 0:
+            self._eager_list = self._lists[0]
+        else:
+            self._lazy_sequence = itertools.chain(*self._lists)
+            """chain of generators
+            this will be consumed over time whenever we need elems.
+            """
+            self._eager_list = None
+            """This is the list where we save indices
+            whenever we generate them from the lazy sequence.
+            """
+            self._offset = offset
+
+        self._known_length = known_length
+
+    def _to_eager(self):
+        if self._eager_list is not None:
+            return
+        self._eager_list = [el + self._offset for el in self._lazy_sequence]
```

Comment:
Maybe we can further optimize using NumPy both as a storage for _eager_list and both to optimize summing the _offset to each index.

Reply:
This is useful if we do it every time we offset (there are many places). Here most cost is due to the full iteration of the iteration. Not sure numpy would help a lot. But in general, converting everything to tensors may speed up general subsets and concat ops.